Add item requests and gateway

SergeyLeshchev opened the pull request.

SergeyLeshchev committed:
Разделить монолитный проект shareit на модули согласно ТЗ №16

1. Добавить модули server и getaway
2. Настроить docker-compose.yaml, Dockerfile, pom.xml, properties
3. Добавить классы Client в getaway
4. Настроить и проверить работу одного метода createUser

SergeyLeshchev committed:
Реализовать модуль gateway, чтобы тесты 15 спринта проходили успешно

1. Добавить в gateway обработку запросов для сущностей User, Item, Booking, Comment
2. Изменить DTO и мапперы, чтобы пользователю возаращались только DTO

SergeyLeshchev committed:
Выполнить ТЗ №16

1. Добавить функционал с сущностью ItemRequest
2. Добавить тесты

SergeyLeshchev committed:
Выполнить ТЗ №16

1. Изменения кодстайл

Ivanov-s-tmn commented on gateway/src/main/java/ru/practicum/shareit/exception/ErrorHandler.java:
Добавь обработку ошибок валидации

Ivanov-s-tmn commented on gateway/src/main/java/ru/practicum/shareit/item/ItemController.java:
```diff
@@ -0,0 +1,65 @@
+    @PostMapping("/{itemId}/comment")
+    public ResponseEntity<Object> createComment(@RequestHeader("X-Sharer-User-Id") Long userId,
+                                                @PathVariable Long itemId,
+                                                @Valid @RequestBody CommentDto commentDto) {
```
Лучше создать отдельную дтошку, нужно только поле текста

Ivanov-s-tmn commented on gateway/src/main/java/ru/practicum/shareit/user/UserController.java:
```diff
@@ -0,0 +1,56 @@
+    private final UserClient userClient;
+
+    @PostMapping
+    public ResponseEntity<Object> createUser(@RequestBody @Valid UserDto userDto) {
```
Дтошки для запросов и ответов лучше разделять, обычно дтошки с ответами более полные

Ivanov-s-tmn commented on gateway/src/main/java/ru/practicum/shareit/user/UserController.java:
```diff
@@ -0,0 +1,56 @@
+
+    @PatchMapping("/{userId}")
+    public ResponseEntity<Object> updateUser(@Positive @PathVariable Long userId,
+                                             @RequestBody UpdateUserRequest updateUserRequest) throws BadRequestException {
```
Зачем `throws BadRequestException`

Ivanov-s-tmn commented on server/pom.xml:
```diff
@@ -0,0 +1,260 @@
+            <artifactId>spring-boot-starter-test</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-validation</artifactId>
+        </dependency>
```
Валидации быть не должно, так что зависимость тоже ну нужна

Ivanov-s-tmn commented on server/src/main/java/ru/practicum/shareit/booking/BookingServiceImpl.java:
```diff
@@ -92,24 +90,24 @@ public Booking getBookingById(Long userId, Long bookingId) {
         userRepository.findById(userId)
                 .orElseThrow(() -> new NotFoundException("Пользователь не найден"));
-        if (status.equals(Status.ALL)) {
+        if (state.equals(BookingState.ALL)) {
             return bookingRepository.findAllByBookerId(userId);
         }
-        return bookingRepository.findAllByBookerIdAndStatus(userId, status);
+        return bookingRepository.findAllByBookerIdAndState(userId, state);
     }
 
     @Override
-    public List<Booking> getAllBookingsByOwner(Long userId, Status status) {
+    public List<Booking> getAllBookingsByOwner(Long userId, BookingState state) {
         // Проверка существования пользователя
         userRepository.findById(userId)
                 .orElseThrow(() -> new NotFoundException("Пользователь не найден"));
-        if (status.equals(Status.ALL)) {
+        if (state.equals(BookingState.ALL)) {
             return bookingRepository.findAllByItemOwnerId(userId);
         }
-        return bookingRepository.findAllByItemOwnerIdAndStatus(userId, status);
+        return bookingRepository.findAllByItemOwnerIdAndState(userId, state);
     }
```
В логике этих методов не все так просто, кажется, я писал об этом в ревью прошлого спринта.

При Future нужно вернуть только будущие бронирования (статуса FUTURE у бронирований нет, тут нужно работать с временем), для PAST - прошлые, при CURRENT - текущие. И все они должны быть с сортировкой от новых к старым

SergeyLeshchev replied:
Я так же поменял интеграционные тесты методов репозитория, а юнит-тесты менять не стал, потому что для сервиса мало что изменилось. Он все так же возвращает список, который в готовом виде получает из репозитория и всё

Ivanov-s-tmn commented on server/src/main/java/ru/practicum/shareit/item/dto/CommentDto.java:
```diff
@@ -0,0 +1,18 @@
+@Data
+public class CommentDto {
+    private Long id;
+    @NotBlank
```
Валидация не нужна на уровне сервера

Ivanov-s-tmn commented on server/src/main/java/ru/practicum/shareit/request/ItemRequestServiceImpl.java:
```diff
@@ -0,0 +1,76 @@
+                .map(ItemRequestMapper::maptoItemRequestOutDto)
+                .peek(ir -> {
+                    // Находим каждому ItemRequestOutDto список вещей
+                    List<ItemResponse> items = itemRepository.findAllByRequestId(ir.getId()).stream()
```
Запросы в цикле

Ivanov-s-tmn commented on server/src/test/java/ru/practicum/shareit/user/UserRepositoryIT.java:
Почему пустой?

SergeyLeshchev committed:
Сделать правки №1 для ТЗ №16

1. Убрать запросы в циклах
2. Изменить логику BookingService методов для поиска списка бронирований со статусом
3. Разделить все DTO на ответные и запросные DTO
4. Убрать валидацию на уровне сервера
5. Добавить обработку ошибок валидации в gateway
6. Добавить интеграционные тесты для сервисов
7. Удалить лишний код

SergeyLeshchev committed:
Исправления чекстайл

SergeyLeshchev committed:
Добавить в CommentRequestDto конструктор без параметров

SergeyLeshchev committed:
Тестировать сервисы в интеграционных тестах изолированно
